### ISSUE TITLE
Some clients block forever when the server is (was) under high pressure

### ISSUE THREAD

**Issue by Kelvin-Ng** [grpc/grpc#12364]
### What version of gRPC and what language are you using?
grpc-1.4.5, C++
 
### What operating system (Linux, Windows, …) and version?
Arch Linux
Linux *** 4.12.8-2-ARCH #1 SMP PREEMPT Fri Aug 18 14:08:02 UTC 2017 x86_64 GNU/Linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 GCC 7.1.1

### What is your CPU?
Intel i5-4570. With 4 physical cores and no hyperthreading.
 
### What did you do?
1. Change the example examples/cpp/helloworld/greeter_async_server.cc to serve incoming requests multithreadedly. In particular:
```diff
diff --git a/examples/cpp/helloworld/greeter_async_server.cc b/examples/cpp/helloworld/greeter_async_server.cc
index e40889a84c..ba8d7c9588 100644
--- a/examples/cpp/helloworld/greeter_async_server.cc
+++ b/examples/cpp/helloworld/greeter_async_server.cc
@@ -61,8 +61,15 @@ class ServerImpl final {
     server_ = builder.BuildAndStart();
     std::cout << "Server listening on " << server_address << std::endl;
 
+    std::vector<std::thread> thrs;
+    thrs.reserve(4);
     // Proceed to the server's main loop.
-    HandleRpcs();
+    for (int i = 0; i < 4; ++i) {
+        thrs.emplace_back(&ServerImpl::HandleRpcs, this);
+    }
+    for (int i = 0; i < 4; ++i) {
+        thrs[i].join();
+    }
   }
```

2. Compile and run the modified greeter_async_server
3. Compile greeter_async_client and run 10000 instances of it *roughly simultaneously* by `for i in {1..10000}; do ./greeter_async_client > /dev/null 2> /dev/null & done`
 
### What did you expect to see?
All the clients finish after a while
 
### What did you see instead?
Some, but not all, of the clients block forever. Neither the server nor the blocking clients consume CPU.

### What did you try afterwards?
I tried to use a mutex to protect around the call to `service_->RequestSayHello`, but the same problem exists.

**Comment by vjpai** [comment]
I have seen a similar report of another problem that was solved by #12648 . Can you repro this problem at current master?

**Comment by vjpai** [comment]
Closing since no response and since the related problem was indeed resolved by #12648
